### ISSUE TITLE
Qlobber as a Redis Module / C lib

### ISSUE THREAD

**Issue by behrad** [davedoesdev/qlobber#10]
I'm thinking of the idea of implementing a qlobber/trie as a http://redismodules.com
Even a C implementation would be more optimized memory/cpu wise as @mcollina suggested to be bound in a node.js process.
What dya think @davedoesdev ?

**Comment by davedoesdev** [comment]
@behrad sounds good. A C version should be a lot quicker, though you'll likely have to pay attention to keep the C code optimized as V8 is pretty good (TurboFan is on the way too).

Will you be making it open source?

**Comment by behrad** [comment]
Sure, if I have enough time to write it... :))

**Comment by davedoesdev** [comment]
This repo will only be interested in a Node version (playing with Redis Modules sounds fun though - let me know how it goes).

Do you think a version of `Qlobber` which uses Node N-API would be any quicker than the pure JS version? @mcollina what do you think? Wouldn't it have to return data using v8 Sets/Maps anyway? How would one customise it easily (I guess by changing the C source).

**Comment by mcollina** [comment]
NaN is faster than N-API. Implementing a Trie in C++ might a) occupy less memory and b) be faster, but on the other side the time to cross the boundary.

**Comment by davedoesdev** [comment]
@behrad I've improved performance by switching qlobber's trie implementation from using Maps to using objects.

Benchmark before:

```
Running "exec:bench" (exec) task

  add_match_remove x20,000
  ┌─────────┬────────────┬──────────────┬──────────┐
  │         │ total (ms) │ average (ns) │ diff (%) │
  ├─────────┼────────────┼──────────────┼──────────┤
  │ default │      2,535 │      126,741 │        - │
  │ dedup   │      2,907 │      145,363 │       15 │
  │ mapval  │      2,937 │      146,832 │       16 │
  └─────────┴────────────┴──────────────┴──────────┘

  match x20,000
  ┌─────────┬────────────┬──────────────┬──────────┐
  │         │ total (ms) │ average (ns) │ diff (%) │
  ├─────────┼────────────┼──────────────┼──────────┤
  │ default │      2,259 │      112,964 │        - │
  │ dedup   │      2,562 │      128,088 │       13 │
  │ mapval  │      2,606 │      130,290 │       15 │
  └─────────┴────────────┴──────────────┴──────────┘

Running "exec:bench-add-many" (exec) task

  add_many x1
  ┌─────────┬────────────┬───────────────┬──────────┐
  │         │ total (ms) │  average (ns) │ diff (%) │
  ├─────────┼────────────┼───────────────┼──────────┤
  │ dedup   │        277 │   277,421,505 │        - │
  │ mapval  │        402 │   401,738,740 │       45 │
  │ default │      5,015 │ 5,014,769,728 │    1,708 │
  └─────────┴────────────┴───────────────┴──────────┘

Running "exec:bench-match-many" (exec) task

  match_many x1
  ┌─────────┬────────────┬───────────────┬──────────┐
  │         │ total (ms) │  average (ns) │ diff (%) │
  ├─────────┼────────────┼───────────────┼──────────┤
  │ default │      4,616 │ 4,615,824,516 │        - │
  │ dedup   │      6,289 │ 6,288,792,523 │       36 │
  │ mapval  │      8,121 │ 8,121,452,216 │       76 │
  └─────────┴────────────┴───────────────┴──────────┘
```

Benchmark after:

```
Running "exec:bench" (exec) task

  add_match_remove x20,000
  ┌─────────┬────────────┬──────────────┬──────────┐
  │         │ total (ms) │ average (ns) │ diff (%) │
  ├─────────┼────────────┼──────────────┼──────────┤
  │ default │      1,769 │       88,459 │        - │
  │ dedup   │      2,131 │      106,557 │       20 │
  │ mapval  │      2,340 │      117,004 │       32 │
  └─────────┴────────────┴──────────────┴──────────┘

  match x20,000
  ┌─────────┬────────────┬──────────────┬──────────┐
  │         │ total (ms) │ average (ns) │ diff (%) │
  ├─────────┼────────────┼──────────────┼──────────┤
  │ default │      1,394 │       69,698 │        - │
  │ mapval  │      1,761 │       88,069 │       26 │
  │ dedup   │      1,837 │       91,835 │       32 │
  └─────────┴────────────┴──────────────┴──────────┘

Running "exec:bench-add-many" (exec) task

  add_many x1
  ┌─────────┬────────────┬───────────────┬──────────┐
  │         │ total (ms) │  average (ns) │ diff (%) │
  ├─────────┼────────────┼───────────────┼──────────┤
  │ dedup   │        413 │   412,949,841 │        - │
  │ mapval  │        541 │   541,028,758 │       31 │
  │ default │      5,270 │ 5,269,747,786 │    1,176 │
  └─────────┴────────────┴───────────────┴──────────┘

Running "exec:bench-match-many" (exec) task

  match_many x1
  ┌─────────┬────────────┬────────────────┬──────────┐
  │         │ total (ms) │   average (ns) │ diff (%) │
  ├─────────┼────────────┼────────────────┼──────────┤
  │ default │      6,689 │  6,688,838,956 │        - │
  │ dedup   │      8,498 │  8,498,396,189 │       27 │
  │ mapval  │     10,486 │ 10,486,112,805 │       57 │
  └─────────┴────────────┴────────────────┴──────────┘
```

**Comment by davedoesdev** [comment]
Published as version 0.9.0.

**Comment by davedoesdev** [comment]
Match is now faster (by at least a third) but add is slower (by the same amount).

**Comment by davedoesdev** [comment]
@behrad I know you had a problem with add so the question is how to make both match and add faster.

**Comment by davedoesdev** [comment]
@mcollina yes the problem is crossing the boundary. When matching you'd want to create a JS Set and return it.

Another question is whether focusing on squeezing out performance from a single process is worth it. If you want scale, perhaps @behrad 's idea of using Redis and scaling out is better when you need more performance in the system as a whole.

**Comment by davedoesdev** [comment]
I've reverted back to Map for version 0.10.0 of qlobber. Map is the right thing to use in theory and hopefully it will get some improvements in lookup speed.

**Comment by behrad** [comment]
https://github.com/mcollina/aedes/issues/123#issuecomment-318260710

**Comment by davedoesdev** [comment]
On the branch `shortcut-cache` I've added a 'shortcut' cache which makes adding values for the same topic much quicker. This brings the add_many benchmark down to:

```
  add_many x1
  ┌─────────┬────────────┬───────────────┬──────────┐
  │         │ total (ms) │  average (ns) │ diff (%) │
  ├─────────┼────────────┼───────────────┼──────────┤
  │ dedup   │        125 │   125,372,442 │        - │
  │ mapval  │        215 │   214,809,251 │       71 │
  │ default │      5,098 │ 5,097,730,099 │    3,966 │
  └─────────┴────────────┴───────────────┴──────────┘
```

At the moment the size of the shortcut cache grows with the number of topics added. I'm considering whether (a) to make the shortcut cache optional and (b) whether to limit the size to a constant (eject LRU topics for instance. Another option would make the caller decide whether to use the cache when calling `add()`.

@behrad @mcollina opinions?

**Comment by davedoesdev** [comment]
The shortcut cache is now in master and published as version 0.11.0.
When creating a `Qlobber`, you can turn on the shortcut cache with the following new option:

`{Boolean} cache_adds` Whether to cache topics when adding topic matchers. This will make adding multiple matchers for the same topic faster at the cost of extra memory usage. Defaults to `false`.

**Comment by davedoesdev** [comment]
Closing: because of boundary crossing, using native hashtable/es6-map is slower.
I won't be making a Redis module as part of this project.
